### PR TITLE
Benchmarks: Add Benchmark - Add CNN model benchmarks.

### PR DESCRIPTION
**Pull request by guoshzhao**
**Major Revision**
- Add CNN model benchmarks, including:

> [
>     'alexnet', 'densenet121', 'densenet169', 'densenet201', 'densenet161', 'googlenet', 'inception_v3', 'mnasnet0_5',
>     'mnasnet0_75', 'mnasnet1_0', 'mnasnet1_3', 'mobilenet_v2', 'mobilenet_v3_large', 'mobilenet_v3_small', 'resnet18',
>     'resnet34', 'resnet50', 'resnet101', 'resnet152', 'resnext50_32x4d', 'resnext101_32x8d', 'wide_resnet50_2',
>     'wide_resnet101_2', 'shufflenet_v2_x0_5', 'shufflenet_v2_x1_0', 'shufflenet_v2_x1_5', 'shufflenet_v2_x2_0',
>     'squeezenet1_0', 'squeezenet1_1', 'vgg11', 'vgg11_bn', 'vgg13', 'vgg13_bn', 'vgg16', 'vgg16_bn', 'vgg19_bn', 'vgg19'
> ]

**Minor Revision**
- Add tests for CNN benchmarks. 
- Add resnet101 as the example of CNN benchmark.

### CODE REVIEW

File: tests/benchmarks/model_benchmarks/test_pytorch_cnn.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from superbench.benchmarks import BenchmarkRegistry, Platform, Framework, BenchmarkType, ReturnCode
+from superbench.benchmarks.model_benchmarks.pytorch_cnn import PytorchCNN
+
+
+@decorator.cuda_test
+@decorator.pytorch_test
+def test_pytorch_cnn():
+    """Test pytorch cnn benchmarks."""
+    for model in [
+        'resnet50', 'resnet101', 'resnet152', 'densenet169', 'densenet201', 'vgg11', 'vgg13', 'vgg16', 'vgg19'
```
Comment (cp5555) [inline tests/benchmarks/model_benchmarks/test_pytorch_cnn.py:16]: Maybe we can add more for this?

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                    if curr_step > self._args.num_warmup:
+                        # Save the step time of every training/inference step, unit is millisecond.
+                        duration.append((end - start) * 1000)
+                    if self._is_finished(curr_step, end):
+                        return duration
+
+
+# Register CNN benchmarks.
+# Reference: https://pytorch.org/vision/stable/models.html
+for model in ['resnet50', 'resnet101', 'resnet152', 'densenet169', 'densenet201', 'vgg11', 'vgg13', 'vgg16', 'vgg19']:
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/pytorch_cnn.py:146]: Maybe we can add more for this?

File: tests/benchmarks/model_benchmarks/test_pytorch_cnn.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline tests/benchmarks/model_benchmarks/test_pytorch_cnn.py:1]: The same comments: we should keep the code coverage for patch is > 80%.

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py (hunk 2 above)
Comment (guoshzhao) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:146]: Do you think we could add all the cnn models?

Comment (cp5555) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py]: Yes, although user may not use it in most of time ~

File: tests/benchmarks/model_benchmarks/test_pytorch_cnn.py (hunk 3 above)
Comment (cp5555) [reply tests/benchmarks/model_benchmarks/test_pytorch_cnn.py:1]: Solved.

Comment (guoshzhao) [reply tests/benchmarks/model_benchmarks/test_pytorch_cnn.py:1]: Done. Currently `93.90% of diff hit (target 80.00%)`

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                        duration.append((end - start) * 1000)
+                    if self._is_finished(curr_step, end):
+                        return duration
+
+
+# Register CNN benchmarks.
+# Reference: https://pytorch.org/vision/stable/models.html
+#            https://github.com/pytorch/vision/tree/master/torchvision/models
+MODELS = [
+    'alexnet', 'densenet121', 'densenet169', 'densenet201', 'densenet161', 'googlenet', 'inception_v3', 'mnasnet0_5',
```
Comment (abuccts) [inline superbench/benchmarks/model_benchmarks/pytorch_cnn.py:148]: get it from `torchvision.models.__dict__` dynamically?

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        self._optimizer_type = Optimizer.SGD
+        self._loss_fn = torch.nn.CrossEntropyLoss()
+
+    def add_parser_arguments(self):
+        """Add the CNN-specified arguments."""
+        super().add_parser_arguments()
+
+        self._parser.add_argument('--model_type', type=str, required=True, help='The cnn benchmark to run.')
+        self._parser.add_argument('--image_size', type=int, default=224, required=False, help='Image size.')
+        self._parser.add_argument('--num_classes', type=int, default=100, required=False, help='Num of class.')
```
Comment (abuccts) [inline superbench/benchmarks/model_benchmarks/pytorch_cnn.py:38]: why 100 is the default num_classes instead of 1000?

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py (hunk 4 above)
Comment (guoshzhao) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:148]: There have some other terms in the torchvision.models.__dict__, like:

> 'video',  'quantization', 'detection', 'segmentation', 'utils'.

And I don't want to use the v0.1's style, 

> models.mnasnet.__all__[1:]

It's a little trick since the list may be changed.
I should we should claim what models superbench can support explicitly.

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py (hunk 5 above)
Comment (guoshzhao) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:38]: 100 or 1000 has no impact for the performance.  We use 100 for all models. If you think 1000 is more suitable, we can change to 1000.

Comment (abuccts) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:38]: since most of these models are originally trained on imagenet, I think 1000 is more reasonable.

Comment (guoshzhao) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:38]: Sure, changed to 1000.

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py (hunk 4 above)
Comment (abuccts) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:148]: then do you need to check `hasattr` before register? since different `torchvision` versions may have different lists of models.

**Review by abuccts — APPROVED** [review]

File: superbench/benchmarks/model_benchmarks/pytorch_cnn.py (hunk 4 above)
Comment (guoshzhao) [reply superbench/benchmarks/model_benchmarks/pytorch_cnn.py:148]: Sure.
